accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,802 @@
+            and the inferred value type of `element` is `V`, where `K` and `V`
+            are the types such that `Map<K, V>` is a superinterface of `S` (the
+            result of constraint matching for `X` and `Y` using the constraint
+            `S <: Map<X, Y>`).
+
+        *   If `S` is `dynamic`, then the inferred set element type of `element`
+            is `dynamic`, the inferred key type of `element` is `dynamic`, and
+            the inferred value type of `element` is `dynamic`. *(We produce both
+            a set element type here, and a key/value pair here and rely on other
+            elements to disambiguate.)*
+
+        *   Otherwise it is an error (either because we cannot disambiguate,
+            such as with something that implements both Map and Iterable, or
+            because it is a spread of a non-spreadable type).
+
+    *   If `P` is `Set<Ps>` then let `S` be the inferred type of `e1` in context
+        `Iterable<Ps>`:
+
+        *   If `S` is a non-`Null` subtype of `Iterable<Object>`, then the
+            inferred set element type of `element` is `T` where `T` is the type
+            such that `Iterable<T>` is a superinterface of `S` (the result of
+            constraint matching for `X` using the constraint `Iterable<X> <:
+            S`).
+
+        *   If `S` is `dynamic`, then the inferred set element type of `element`
+            is `dynamic`.
+
+        *   Otherwise it is an error.
+
+    *   If `P` is `Map<Pk, Pv>` then let `S` be the inferred type of `e1` in
+        context `P`:
+
+        *   If `S` is a non-`Null` subtype of `Map<Object, Object>`, then the
+            inferred key type of `element` is `K` and the inferred value type of
+            `element` is `V`, where `K` and `V` are the types such that `Map<K,
+            V>` is a superinterface of `S` (the result of constraint matching
+            for `X` and `Y` using the constraint `Map<X, Y> <: S`).
+
+        *   If `S` is `dynamic`, then the inferred key type of `element` is
+            `dynamic`, and the inferred value type of `element` is `dynamic`.
+
+        *   Otherwise it is an error.
+
+*   If `element` is a `ifElement` with one `element`, `p1`, and no `else`:
+
+    The condition is inferred with a context type of `bool`.
+
+    *   If the inferred set element type of `p1` is `S` then the inferred set
+        element type of `element` is `S`.
+
+    *   If the inferred key type of `p1` is `K` and the inferred value type of
+        `p1` is `V` then the inferred key and value types of `element` are `K`
+        and `V`.
+
+    *Note that both of the above cases can simultaneously apply because of
+    `dynamic` spreads.*
+
+*   If `element` is an `ifElement` with two `element`s, `p1` and `p2`:
+
+    The condition is inferred with a context type of `bool`.
+
+    It is a compile error if `p1` has an inferred set element type and `p2` does
+    not, or if `p2` has an inferred set element type and `p1` does not. *In
```

Comment:
If you take the approach where `...?e` where `e` has type `Null` generates neither a set element type nor a key/value pair, then just change this to make it an error whenever one of the parts has a set element type and no key/value pair and the other part has a key/value pair and no set element type.

Reply:
Done, by way of defining "must be a set" and "must be a map".